backup: add resume and onfail/cancel logic to compactions

kev-cao opened the pull request:
Compactions were jobified in #141183, but missing support for job resumption and cancellation handling. This commit adds support for handling those cases.

An additional PR should be made to add telemetry for compactions.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141753)
<!-- Reviewable:end -->

kev-cao committed:
backup: add resume and onfail/cancel logic to compactions

Compactions were jobified in #141183, but missing support for job
resumption and cancellation handling. This commit adds support
for handling those cases.

An additional PR should be made to add telemetry for
compactions.

Epic: none

Release note: None

kev-cao commented on pkg/backup/backup_compaction_test.go:
```diff
@@ -270,6 +271,48 @@ crdb_internal.json_to_pb(
+		jobutils.WaitForJobToPause(t, db, jobID)
+		db.Exec(t, "RESUME JOB $1", jobID)
+		waitForSuccessfulJob(t, tc, jobID)
+		validateCompactedBackupForTables(t, db, []string{"foo", "bar"}, "'nodelocal://1/backup'")
```
I've realized in my testing that this lets through a couple of false negatives. `validateCompactedBackupForTables` simply restores from the latest backup. If the tests were created in isolation, this might not be too big of an issue, but since they all share the same cluster, one test that may have failed to create a compaction would actually just restore from a previous test.

I think we can fix this by having `validateCompactedBackupForTables` take in the `start` and `end` of the compaction, and it should also validate that a backup spanning that timespan exists. But I'll do this in another PR.

msbutler replied:
could you instead use a different collection dir for each subtest instead of `nodelocal://1/backup` for everything?

kev-cao replied:
Oh yeah that would do the trick. I do think having an extra check that makes sure that there exists a backup that spans the entire time from `start` to `end` is good though.

kev-cao replied:
I'll do this in a separate PR though

msbutler commented on pkg/backup/backup_compaction.go:
```diff
@@ -92,128 +95,273 @@ func StartCompactionJob(
 }
 
-func compactIncrementals(
+func (b *backupResumer) OnFailOrCancelCompaction(
```
nit: i took a closer look at the og onfailorcancel routine and it actually noops if there's no pts on the job record, so I actually think the og routine is totally fine. I think you can delete this. 

We'll worry about metrics/telemetry later.

msbutler commented on pkg/backup/backupencryption/encryption.go:
```diff
@@ -338,12 +338,9 @@ func GetEncryptionFromBase(
 	makeCloudStorage cloud.ExternalStorageFromURIFactory,
 	baseBackupURI string,
-	encryptionParams jobspb.BackupEncryptionOptions,
+	encryptionParams *jobspb.BackupEncryptionOptions,
```
this kind of small refactor should be in a separate commit. 

I'm also not sure what it is seeking to accomplish: previously we could  avoid `makeCloudStorage` call if there was no encryption, but now we have to make the call regardless?

kev-cao replied:
Oh this issue was caused by the fact that `initialDetails` would start off with a non-nil `encryption` because it was always passed an encryption from the builtin. But through our flow of creating the updated details, if we aren't using encryption, encryption actually becomes `nil` in the updated details (from `GetEncryptionFromBaseStore`).

So on resume, `initialDetails.EncryptionOptions` is `nil`, which led to some `nil` pointer exceptions because some of our code expects a `*jobspb.BackupEncryptionOptions` and other parts expect a `jobspb.BackupEncryptionOptions`. Classic backup happens to avoid this because on resume, we skip all of this work. But with compaction, since we have to validate the compaction chain, we pull all of the manifests for the backup chain, and as a consequence, end up requiring the encryption from `initialDetails`.

msbutler commented on pkg/backup/backup_compaction.go:
```diff
@@ -554,73 +767,151 @@ func resolveBackupDirs(
+	return &m, nil
+}
+
+func getRestoreChain(
```
nit: i think we should use the term backup chain, not restore chain.

msbutler reviewed (commented):
I took a pass at the production code. i'll take a look at that tests later

kev-cao commented on pkg/backup/backup_compaction.go:
```diff
@@ -92,128 +95,273 @@ func StartCompactionJob(
+	// resolution and writing of backup lock due to the need to verify that the
+	// compaction chain is a valid chain. Naturally, the details can only be updated
+	// once the backup destination has been resolved.
+	compactChain, updatedDetails, err := getCompactionChainAndUpdatedDetails(
```
Oh yeah that's a good point. Updated!

kev-cao commented on pkg/backup/backup_compaction.go:
```diff
@@ -92,128 +95,273 @@ func StartCompactionJob(
+		// TODO (kev-cao): Add telemetry for backup compactions.
+	}
+
+	// For all backups, partitioned or not, the main BACKUP manifest is stored at
```
Yeah I agree, after splitting this up instead of integrating it early, I found some areas where improvements could be made because the code was now split up.

msbutler commented on pkg/backup/backup_compaction_test.go:
```diff
@@ -270,6 +271,48 @@ crdb_internal.json_to_pb(
+		db.Exec(t, "INSERT INTO foo VALUES (2, 2)")
+		db.Exec(t, incBackupCmd)
+		db.Exec(t, "CREATE TABLE bar (a INT, b INT)")
+		db.Exec(t, "INSERT INTO bar VALUES (1, 1)")
```
why does this test require 4 inc backups?

kev-cao replied:
Oh I suppose it doesn't, I've gotten into the habit of trying to make sure there are "meaningful" compactions, but I suppose the other tests cover that so there isn't really a need for it in this test.

msbutler commented on pkg/backup/backup_compaction_test.go:
```diff
@@ -270,6 +271,48 @@ crdb_internal.json_to_pb(
+
+		db.Exec(t, "SET CLUSTER SETTING jobs.debug.pausepoints = ''")
+		db.Exec(t, "INSERT INTO bar VALUES (4, 4)")
+		db.Exec(t, incBackupCmd)
```
is this extra vanilla inc backup necessary?

kev-cao replied:
Nope, had it for the same reason above.

msbutler reviewed (commented):
looking better!

kev-cao commented on pkg/backup/backup_compaction.go:
```diff
@@ -92,128 +95,257 @@ func StartCompactionJob(
+	// Resolve the backup destination. If we have already resolved and persisted
+	// the destination during a previous resumption of this job, we can re-use
+	// the previous resolution.
+	if needsInitializing {
```
Oh right, the dependency on backupDest got moved into the block so this can be moved too now. That's awesome, I hated this extra block.

kev-cao commented on pkg/backup/backup_compaction.go:
```diff
@@ -445,21 +636,24 @@ func openSSTs(
 		return jobspb.BackupDetails{}, errors.New("no backup manifests to compact")
 	}
+	// If we have already resolved the updated details during a previous
+	// resumption of this job, we can re-use the initial details.
```
Oh yeah good catch, not necessary anymore

kev-cao commented:
TFTR!

Bors r=msbutler
